File: java/binaryTree/PreorderTraversal.java
/**
 * Given the root of a binary tree, return the preorder traversal of its nodes' values.
 * 
 * Example 1:
 * Input: root = [1,null,2,3]
 * Output: [1,2,3]
 * 
 * Constraints:
 * The number of nodes in the tree is in the range [0, 100].
 * -100 <= Node.val <= 100
 */

package binaryTree;

import java.util.LinkedList;
import java.util.List;
import java.util.Stack;

public class PreorderTraversal {

    public static void main(String[] args) {
        /*
         *            5
         *           / \
         *          2   4
         *         /   /
         *        1   3
         */
        TreeNode node1 = new TreeNode(1, null, null);
        TreeNode node2 = new TreeNode(2, node1, null);
        TreeNode node3 = new TreeNode(3, null, null);
        TreeNode node4 = new TreeNode(4, node3, null);
        TreeNode node5 = new TreeNode(5, node2, node4);
        PreorderTraversal p = new PreorderTraversal();
        for (Integer val : p.preorderTraversal(node5)) {
            System.out.print(val + " ");
        }
        System.out.println();
        for (Integer val : p.preorderTraversal2(node5)) {
            System.out.print(val + " ");
        }
        System.out.println();
    }
    
    /**
     * 递归法：深度优先遍历
     */
    public List<Integer> preorderTraversal(TreeNode root) {
        LinkedList<Integer> list = new LinkedList<Integer>();
        this.dfs(root, list);
        return list;
    }
    
    private void dfs(TreeNode node, LinkedList<Integer> list) {
        if (node == null) {
            return;
        }
        list.add(node.val);
        dfs(node.left, list);
        dfs(node.right, list);
    }
    
    /**
     * 迭代法：先将当前节点加入到数组中，然后迭代访问左子树直到最左边的叶子节点，最后迭代访问右子树，
     * 访问方法和之前一样
     */
    public List<Integer> preorderTraversal2(TreeNode root) {
        LinkedList<Integer> list = new LinkedList<Integer>();
        Stack<TreeNode> stack = new Stack<TreeNode>();
        TreeNode cur = root;
        while (cur != null || !stack.isEmpty()) {
            if (cur != null) {
                list.add(cur.val);
                if (cur.right != null) {
                    stack.push(cur.right);
                }
                cur = cur.left;
            } else {
                cur = stack.pop();
            }
        }
        return list;
    }

}

File: javascript/binaryTree/PreorderTraversal.js
/**
 * 144. Binary Tree Preorder Traversal
 *
 * Given the root of a binary tree, return the preorder traversal of its nodes' values.
 *
 * Example 1:
 * Input: root = [1,null,2,3]
 * Output: [1,2,3]
 *
 * Constraints:
 * The number of nodes in the tree is in the range [0, 100].
 * -100 <= Node.val <= 100
 */
import { TreeNode } from './TreeNode.js';
import { createTree } from './LevelOrder.js';

/**
 * 深度优先遍历
 */
var preorderTraversal = function(root) {
    let list = [];
    preorder(root, list);
    return list;
}

var preorder = function(node, list) {
    if (!node) {
        return;
    }
    list.push(node.val);
    preorder(node.left, list);
    preorder(node.right, list);
}

/**
 * 先将当前节点加入到结果中，然后迭代访问左子树直到最左边的叶子节点，最后迭代访问右子树，
 * 访问方法和之前一样
 */
var preorderTraversal2 = function(root) {
    let list = [];
    let stack = [];
    while (root || stack.length) {
        while (root) {
            list.push(root.val);
            stack.push(root);
            root = root.left;
        }
        root = stack.pop();
        root = root.right;
    }
    return list;
}

var main = function() {
    let root = createTree();
    console.log('preorder: ' + preorderTraversal(root));
    console.log('preorder: ' + preorderTraversal2(root));
}

main();
